### DISCUSSION TITLE
Improved i18n support

### DISCUSSION THREAD

**Discussion by nbouvrette** [vercel/next.js#20240] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

I was reading the Next.js documentation and there was a mention of "Internationalized Routing is the first in a series of features that will make internationalizing and localizing your project easier"

I have a few ideas I would like to share from my personal i18n experience with other frameworks, and also considering myself new to Next.js.

## Describe the solution you'd like

Few ideas:

- for sub-path routing, would it be possible to also display the locale for the default one - it's easy to presume that English is `en-US` while many countries speak English and have different ways to format numbers and dates. Making it explicit removes this confusion.

- for the home page, instead of redirecting, maybe it would be possible to display the preferred language on `/` directly with the proper `hreflang` meta tags, which brings me to my next topic:

- page variants - a lot of sites (e.g. Apple) are failing to map page (language) variants when in fact it should be part of an easy to use i18n router. When you switch the language you should be able to re-render the page (if available) with your new locale. This should also enable the automatic generation of `hreflang` meta tags. Fallback should be to the root page of this locale only which brings me to my next topic:

- localizing URLs - it's great to have the locale suffix but when I see an English URL after selecting my non-English language, this seems like a half-finished job. Of course, localized URLs should also support UTF-8 characters which have been proven to also have positive impacts on customer trust. Of course, to do this correctly, the URLs themselves would need to be added in part of the strings file to make them easy to translate. It would make sense that each page has its own localizable strings that would both include the URL and the header metadata. 

- Out of the box string display support - while `react-i18next` seems quite popular, it's doing two things that I feel are sub-optimal:

	- Uses JSON files: this prevents from adding inline-context to translators since JSON does not support comments
	- Uses folders to separate languages - when you have a big project, managing multiple folders becomes quite difficult. Ideally, each page should have its own string files and all use unique keys (never use the string itself since it can cause problems during translation because of the translation memory and frequent collisions).

	
Some ideas for an ideal out of the box solution:

- a file format (e.g. `.properties`) that makes it simple for localization (simple key and value, supports inline comments). The file format should also exist and be well supported by translation tools. The main two file formats I can think of that would fit this would be `.properties` and `.yaml` (key/value only) files. The benefit of `.properties` file is its good integration with IntelliJ which makes them easy to use when you support many locales (IntelliJ hides all the files and make it easy to compare keys and values across languages).
- good IDE integration (the `.properties` files with IntelliJ is fantastic - I could not find similar integration with Visual Studio Code)
- supports standard i18n (Unicode CLDR) formats for numbers, dates, plural, etc. (e.g. MessageFormat)

## Describe alternatives you've considered

No alternatives found so far, other than creating my own solution?

## Additional context

This type of out of the box feature seems inline with Next.js's philosophy and could give Next.js an edge over its competitors.

**Comment by iduuck** [comment]
> for sub-path routing, would it be possible to also display the locale for the default one - it's easy to presume that English is en-US while many countries speak English and have different ways to format numbers and dates. Making it explicit removes this confusion.

Agreeing with you on this one. It's been a pain, since I assumed, that each locale has it's own sub-path, which would make it much more clearer (also in sight on UX) to the user, which locale is currently visible (also no unnecessary redirects from `/` to `/de` for example because users are indeed visiting the "en" version).

@Next.js Team: Please implement the usage of sub-paths for each locale! 🙏

**Comment by nbouvrette** [comment]
We just released a new package that overcomes this issue (and many others): https://github.com/Avansai/next-multilingual

Looking forward to hearing feedback on our approach.
